elfpy/utils/apeworx_integrations.py

```diff
@@ -0,0 +1,250 @@
+    hyperdrive_contract: ape.contracts.base.ContractInstance
+        Ape project `ContractInstance
+        <https://docs.apeworx.io/ape/stable/methoddocs/contracts.html#ape.contracts.base.ContractInstance>`_
+        wrapped around the initialized MockHyperdriveTestnet smart contract.
+    agent_address: ape.api.accounts.AccountAPI
+        Ape address container, or `AccountAPI
+        <https://docs.apeworx.io/ape/stable/methoddocs/api.html#ape.api.accounts.AccountAPI>`_
+        representing the agent which is executing the action.
+    base_amount: int
+        Unsigned int-256 representation of the base amount that the agent wishes to provide.
+
+    Returns
+    -------
+    Tuple[dict[str, Any], ape_ethereum.transactions.Receipt]
+        A tuple containing an update dictionary for the Hyperdrive pool state
+        as well as the Ape-ethereum transaction `receipt
+        <https://docs.apeworx.io/ape/stable/methoddocs/api.html#ape.api.transactions.ReceiptAPI>`_.
+    """
+    with ape.accounts.use_sender(agent_address):  # sender for contract calls
+        if trade_prefix == AssetIdPrefix.LONG:  # open a long
+            min_output = 0
+            as_underlying = False  # mockHyperdriveTestNet does not support as_underlying=True
+            tx_receipt = hyperdrive_contract.openLong(  # type: ignore
+                trade_amount,  # base
+                min_output,
+                agent_address,
+                as_underlying,
+            )
+        elif trade_prefix == AssetIdPrefix.SHORT:  # open a short
+            max_deposit = trade_amount
+            as_underlying = False  # mockHyperdriveTestNet does not support as_underlying=True
+            tx_receipt = hyperdrive_contract.openShort(  # type: ignore
+                trade_amount,  # bonds
+                max_deposit,
+                agent_address,
+                as_underlying,
+            )
+        else:
+            raise ValueError(f"{trade_prefix=} must be 0 (long) or 1 (short).")
+        # Return the updated pool state & transaction result
+        transfer_single_event = get_transaction_trade_event(tx_receipt)
+        # The ID is a concatenation of the current share price and the maturity time of the trade
+        token_id = int(transfer_single_event["id"])
+        prefix, maturity_timestamp = decode_asset_id(token_id)
+        pool_state = hyperdrive_contract.getPoolInfo().__dict__  # type: ignore
+        pool_state["block_number_"] = tx_receipt.block_number
+        pool_state["token_id_"] = token_id
+        pool_state["prefix_"] = prefix
+        pool_state["maturity_timestamp_"] = maturity_timestamp
+        logging.debug("hyperdrive_pool_state=%s", pool_state)
+    return pool_state, tx_receipt
+
+
+def ape_close_position(
+    trade_prefix: AssetIdPrefix,
+    hyperdrive_contract: ContractInstance,
+    agent_address: AccountAPI,
+    bond_amount: int,
+    maturity_time: int,
```

Comment:
is this in seconds or years?  add to the comment, its gonna be hard to keep track :P